### ISSUE TITLE
Parrallel consumer stuck in infinite rebalance

### ISSUE THREAD

**Issue by bmaggi** [confluentinc/parallel-consumer#637]
Hi,

I have a quite complexe issue with the 2 latest versions 0.5.2.6 and 0.5.2.7 while everything works correctly in 0.5.2.5. 

When the cpu increase we add machine to distribute the load between them. This obviously trigger a rebalance.
With 0.5.2.6 or 0.5.2.7, the rebalance is so long (> 5 min) that something seems to be blocking a correct rebalance.

I don't see anything particularly meaningful in kafka consumer logs

But I see a lot of _"isPoolQueueLow(): Max loading factor steps reached: 100/100"_ and _"Error committing offsets : Offset commit cannot be completed since the consumer is undergoing a rebalance for auto partition assignment. You can try completing the rebalance by calling poll() and then retry the operation."_ which are quite expected since the rebalance is stuck. 

Fixing the number of machine and restarting solve the issue so I was expecting the fix of https://github.com/confluentinc/parallel-consumer/issues/606 to solve this but unfortunally this was not the case. 

I don't now which logs or metrics may be of use to solve or explain this. 
Feel free to ask for more informations.

Regards
Benoît

**Comment by sangreal** [comment]
I am interesting in the issue, could u please share the trace log of parallel consumer?

**Comment by bmaggi** [comment]
Hi,

Here are the logs of parallel-consumer version 0.5.2.7 in case of 
* a bump of 3 hosts creating a rebalance. 
* rebalance stuck 
* decrease of 3 hosts 

I removed the warn logs, huge spam of "_isPoolQueueLow(): Max loading factor steps reached: 100/100_" which means to me that parallel-consumer is stuck. 

And all _Error committing offsets_ are related to _Offset commit cannot be completed since the consumer is undergoing a rebalance for auto partition assignment. You can try completing the rebalance by calling poll() and then retry the operation._

[extract-parallel-consumer-0527.csv](https://github.com/confluentinc/parallel-consumer/files/12557574/extract-parallel-consumer-0527.csv)

**Comment by sangreal** [comment]
#606 is to solve the issue for reassigning the same partition to the same instance without revoking.
Your case is probably another new issue. Do you have trace level log for parallel consumer turned on?

**Comment by colinkuo** [comment]
This could be related to #402

**Comment by rkolesnev** [comment]
Hi @bmaggi, 

Can you please provide your configs for PC - like ordering, commit strategy, partition assignment strategy etc. as well as how long roughly the processing function takes? 

If you have a reproducible example handy - that would be even better :) 

As far as i can tell since 0.5.2.5 - the only change in commit handling is #548, but - i do not see how it may cause this issue - it only adds additional synchronisation to prevent deadlock between transactional commit and partition revocation.

**Comment by rkolesnev** [comment]
Hi,

Anyone else encountering this issue or have reproducible example or at least configuration of PC and steps leading to this? 
I cannot seem to get to trigger this in tests.

**Comment by pleszczy** [comment]
We've observed this issue in the last two versions. At the moment, we've put in place a health check to restart pods that experience stuck parallel-consume. However, would greatly appreciate a permanent solution.

**Comment by rkolesnev** [comment]
Hi, @pleszczy

Can you please provide details of your setup, configuration, steps to recreate?
Kafka Broker version
PC Version and config - i.e. Ordering, is Producer used?, transactions enabled / disabled, anything else you feel might be relevant?
I could not reproduce in test environment so far.

**Comment by rkolesnev** [comment]
I have recreated the issue with producer set but not used - like in the #648 issue - can you please check if you have same setup - producer is set but not used ?

**Comment by bmaggi** [comment]
> Hi @bmaggi,
> 

Sorry for the delay 

> Can you please provide your configs for PC - like ordering, commit strategy, partition assignment strategy etc. as well as how long roughly the processing function takes?
ParallelConsumerOptions:
* ordering : UNORDERED
* maxConcurrency : 256
* commitMode : PERIODIC_CONSUMER_ASYNCHRONOUS
* batchSize : 1
* thresholdForTimeSpendInQueueWarning : 120 s

* consumer
```
       "enable.auto.commit" = false
        "max.poll.interval.ms" = 600000
        "session.timeout.ms" = 120000
        "max.poll.records" = 1000
        "partition.assignment.strategy" = org.apache.kafka.clients.consumer.CooperativeStickyAssignor

```
* producer
```
        "batch.size" = 65536
        "acks" = all
        "enable.idempotence" = true
        "max.in.flight.requests.per.connection" = 1
        "linger.ms" = 100
        "max.block.ms" = 3600000
        "compression.type" = zstd
        "max.request.size" = 5242880
```

Processing time : Quite fast, most of the time is taken by deserializing the avro data 
after that it's just a redis call and if the data is not present write it in a topic. 

=> @rkolesnev : **so my problem may also be the one in** #648 since we may have quite some time before writing data that is not already present in the redis.  I will have to retry my use case after the fix _for producer set but not used_

> 
> If you have a reproducible example handy - that would be even better :)
> 
Unfortunately, I don't have a simple reproducible example. The issue is only present with a heavy load of data.
I will look if an example may be extracted.  

> As far as i can tell since 0.5.2.5 - the only change in commit handling is #548, but - i do not see how it may cause this issue - it only adds additional synchronisation to prevent deadlock between transactional commit and partition revocation.

**Comment by rkolesnev** [comment]
@bmaggi - yes i think it may be related.
The fix to commit deadlock in #548 has introduced a bug into the logic when producer is not transactional and not used or not used for some period of time. 
I hope to work on a fix for it shortly. 
In a nutshell - the logic there is accessing ProducerManager as committer directly instead of using Committer set on ParallelConsumer. 
Either it needs to validate if Producer is in transactional mode etc - the check there made more robust - or alternatively issue commit through Committer set instead of assuming ProducerManager as committer. As it is now the commit lock is taken but not cleared.
